feat: simplified user administration

NanamiNakano opened the pull request.

NanamiNakano committed:
chore: code cleanup

NanamiNakano committed:
chore: code cleanup

NanamiNakano committed:
feat(user_utils)!: manage user purely from a CSV file

NanamiNakano committed:
feat(user_utils)!: manage user purely from a CSV file

NanamiNakano committed:
feat(user_utils): delete user through csv

NanamiNakano committed:
feat(user_utils): delete user through csv

NanamiNakano committed:
chore: vocabulary

NanamiNakano committed:
chore: vocabulary

NanamiNakano committed:
docs: user administration

NanamiNakano committed:
docs: user administration

NanamiNakano committed:
docs: user administration

Closes: #18

NanamiNakano committed:
docs: user administration

Closes: #18

NanamiNakano committed:
feat:(database): version control

NanamiNakano committed:
feat:(database): version control

NanamiNakano committed:
feat: mercury version

NanamiNakano committed:
feat: mercury version

NanamiNakano committed:
fix: config key is not unique

NanamiNakano committed:
fix: config key is not unique

NanamiNakano committed:
feat: migrate to database with version control

NanamiNakano committed:
feat: migrate to database with version control

NanamiNakano committed:
chore: make code more readable

NanamiNakano committed:
chore: make code more readable

NanamiNakano committed:
fix:: drop old table

NanamiNakano committed:
fix:: drop old table

forrestbao commented:
Can we make user administration as simple as below? 

1. Add a dummy, empty CSV file with the correct columns. 
2. To add, delete, or update users, just edit the CSV file. 
3. When done, use one command to convert CSV to SQLite file. 

The script for user administration shall be called `use_admin.py` It shall have to subcommands: `csv2sqlite` and `sqlite2csv`. 

Tell the admin to put the CSV file in a safe place.

NanamiNakano commented:
I think I've simplified this far, except for not renaming the file. I added the documentation for the new script in the Readme. Or am I not understanding what you are saying?

forrestbao commented:
For all actions other than creating a new user, there should be a one-to-one correspondence between the CSV and the SQLite DB. Thus, the content in CSV shall be the same as the content in the SQLite DB. 

As an example, to delete a user, simply remove the row in the CSV file corresponding to that row and import (called "apply" here) to the SQLite folder. 

Let me work on user admin. I feel mixing command line options and CSV file makes it complicated.

forrestbao committed:
user admin without CSV

forrestbao committed:
user admin without CSV

forrestbao commented:
@NanamiNakano  I have refactored the code to avoid using the CSV as an importing source. The user admin script has been renamed from `user_utils.py` to `user_admin.py`. Please see `user_admin.md` for the usage and test all options, and fix bugs if needed. 

At one place, I used an ugly implementation. Please find it by searching `FIXME` and fix it. 

Third, to avoid confusions, we shall rename `user_name` to `name`. Please make the change throughout the codebase and docs. 

When done, please submit a review.

forrestbao committed:
clean up migration scripts

forrestbao committed:
clean up migration scripts

NanamiNakano committed:
chore: fix typo

NanamiNakano committed:
Merge branch 'feat/database-version-control' into feat/user-management

# Conflicts:
#	user_admin.py

NanamiNakano committed:
fix: user admin

NanamiNakano commented:
I rewrote that part of the code.

For renaming, I think this introduces a break change. Should we do it after we've merged all the existing changes?
